### DISCUSSION TITLE
NestJS cannot resolve SqlEntityManager

### DISCUSSION THREAD

**Discussion by zgid123** [mikro-orm/nestjs#93] [Q&A] [answered]
Hi,

I am trying to use `createQueryBuilder` for MikroORM. This function is from `EntityManager` of package `@mikro-orm/postgresql`, so whenever I inject the entity manager from that package, NestJS cannot resolve.

Here my config:

```ts
// mikro config
import { MikroORMOptions } from '@mikro-orm/core';

import { MikroNamingStrategy } from './config/mikro';

const config: Partial<MikroORMOptions> = {
  type: 'postgresql',
  host: process.env.DB_HOST,
  user: process.env.DB_USER,
  entities: ['./dist/db/models'],
  entitiesTs: ['./src/db/models'],
  password: process.env.DB_PASSWORD,
  namingStrategy: MikroNamingStrategy,
  port: parseInt(process.env.DB_PORT, 10),
  debug: process.env.NODE_ENV === 'development',
  dbName: process.env.DB_NAME || 'ecommerce_development',
  seeder: {
    glob: '!(*.d).{js,ts}',
    path: './dist/db/seeds',
    pathTs: './src/db/seeds',
  },
  migrations: {
    path: './dist/db/migrations',
    pathTs: './src/db/migrations',
  },
};

export default config;
```

```ts
// app.module.ts
import { Module } from '@nestjs/common';
import { MikroOrmModule } from '@mikro-orm/nestjs';

@Module({
  imports: [
    MikroOrmModule.forRoot(),
    ...
  ],
  ...
})
export class AppModule {}
```

```ts
// taxons.module.ts
import { Module } from '@nestjs/common';
import { MikroOrmModule } from '@mikro-orm/nestjs';

import { Taxon } from 'db/models/Taxon';

import { TaxonsController } from './taxons.controller';
import { FilterTaxonsService } from './services/FilterTaxonsService';

@Module({
  imports: [MikroOrmModule.forFeature([Taxon])],
  controllers: [TaxonsController],
  providers: [
    FilterTaxonsService,
    ...
  ],
})
export class TaxonsModule {}
```

```ts
// FilterTaxonsService.ts
import { Injectable } from '@nestjs/common';
import { InjectRepository } from '@mikro-orm/nestjs';
import { EntityManager } from '@mikro-orm/postgresql';
// import { EntityManager } from '@mikro-orm/core';

import type { EntityRepository, FilterQuery } from '@mikro-orm/core';

import { Taxon } from 'db/models/Taxon';
import { FilterService } from 'services/generics/FilterService';

import type {
  IPagyParams,
  IMetadataProps,
} from 'services/generics/FilterService';

interface IFilterParams extends IPagyParams {
  name?: string;
}

@Injectable()
export class FilterTaxonsService extends FilterService {
  constructor(
    @InjectRepository(Taxon)
    private readonly taxonRepository: EntityRepository<Taxon>,
    private readonly em: EntityManager,
  ) {
    super();
  }

  public async exec({
    name,
    page,
    perPage,
  }: IFilterParams): Promise<[Taxon[], IMetadataProps]> {
    let filter: FilterQuery<Taxon> = {};

    if (name) {
      filter = {
        ...filter,
        name: {
          $ilike: name,
        },
      };
    }

    const { limit, offset, ...pagyOptions } = this.pagy({
      page,
      perPage,
    });

    const [taxons, totalRecords] = await this.taxonRepository.findAndCount(
      filter,
      {
        limit,
        offset,
        populate: ['parent'],
      },
    );

    return [
      taxons,
      {
        ...pagyOptions,
        totalRecords,
      },
    ];
  }
}
```

**Comment by zgid123** [comment]
if I use `EntityManager` from package `@mikro-orm/core`, it will work without problem, but that class does not have function `createQueryBuilder`. When I switch to `@mikro-orm/postgresql`, I can use that function, but NestJS cannot resolve it.

I cannot find a solution from document

**Comment by zgid123** [comment]
@B4nan sorry for bothering you. But it seems this problem occurs because the `SqlEntityManager` of `knex` is not the same as `SqlEntityManager` of `postgresql`. So even we have this line of code, we cannot inject `EntityManager` from `@mikro-orm/postgresql` for nestjs.

```
exports: [
  ...(knex ? (contextName ? [] : [knex.SqlEntityManager as any]) : []),
]
```

So I think we should have an option for `forRoot` and `forRootAsync` to specify the driver we are using for our project to import like the knex and mongodb.

For example we do something like

```ts
  static async forRoot(options?: MikroOrmModuleSyncOptions): Promise<DynamicModule> {
    const contextName = this.setContextName(options?.contextName);
    const knex = await tryRequire(() => import('@mikro-orm/knex'));
    const mongo = await tryRequire(() => import('@mikro-orm/mongodb'));
    const driver = (['mongodb', 'knex'].includes(options.driver) || !options.driver) ? null : await tryRequire(() => import(`@mikro-orm/${options.driver}`));

    return {
      module: MikroOrmCoreModule,
      providers: [
        { provide: MIKRO_ORM_MODULE_OPTIONS, useValue: options || {} },
        createMikroOrmProvider(contextName),
        createEntityManagerProvider(options?.scope, EntityManager, contextName),
        ...(knex ? [createEntityManagerProvider(options?.scope, knex.SqlEntityManager, contextName)] : []),
        ...(mongo ? [createEntityManagerProvider(options?.scope, mongo.MongoEntityManager, contextName)] : []),
        ...(driver ? [createEntityManagerProvider(options?.scope, driver.SqlEntityManager, contextName)] : []),
      ],
      exports: [
        contextName ? getMikroORMToken(contextName) : MikroORM,
        contextName ? getEntityManagerToken(contextName) : EntityManager,
        ...(knex ? (contextName ? [] : [knex.SqlEntityManager as any]) : []),
        ...(mongo ? (contextName ? [] : [mongo.MongoEntityManager as any]) : []),
        ...(driver ? (contextName ? [] : [driver.SqlEntityManager as any]) : []),
      ],
    };
  }
```

**Reply by zgid123** [reply]
if you have other approach, lemme know. If the approach above is good for you, I will create a PR

**Reply by B4nan** [reply]
> because the SqlEntityManager of knex is not the same as SqlEntityManager of postgresql.

If this happens, it means you probably have a dependency mismatch, in other words, you have multiple installs of the knex package. Because it is literally a [re-export from knex package](https://github.com/mikro-orm/mikro-orm/blob/master/packages/postgresql/src/index.ts#L1). I'd say this is something to fix in your codebase, not in this package.

**Reply by B4nan** [reply]
Ideally, try to create a complete repro I can take a look myself. Be sure to verify you have the same ORM package versions in the your package.json (except the nestjs adapter) and wipe lockfile/node_modules and reinstall. That usually helps. I dont think what you provided in the OP is enough to reproduce, the problem is not really in code, rather in your dependencies.

What package manager are you using? I also recall issues with pnpm, although those should be now resolved in latest versions.

**Reply by zgid123** [reply]
I am using `pnpm` monorepo. I will create a minimal project and send it for you later.

I checked the package.json, and only one package.json install those packages

```
    "@mikro-orm/core": "^5.5.1",
    "@mikro-orm/migrations": "^5.5.1",
    "@mikro-orm/nestjs": "^5.1.2",
    "@mikro-orm/postgresql": "^5.5.1",
    "@mikro-orm/seeder": "^5.5.1",
```

Maybe I should remove the node_modules and try again. Thanks for your rely

**Reply by B4nan** [reply]
Lastly, you proposed adding `driver` option - it is already there, and it is the actual driver class, not a package name. 

https://mikro-orm.io/docs/configuration#driver

If we should add some escape hack, there is no sense in making it a string and doing dynamic import again (which is the pain point that breaks), instead, we could pass the EM token directly. But I really don't like that solution, it won't help with other symbols (the same issue is there with the new `MikroORM` class extensions added in v5.6) and it is weird to explain at best.

**Reply by B4nan** [reply]
> pnpm monorepo

Now things make more sense :] pnpm is quite opinionated, the only package manager that actually causes mystery issues like this one. 

>     "@mikro-orm/core": "^5.5.1",

I'd try pinning the version and updating explicitly to latest, so `"@mikro-orm/core": "5.6.0"`. The ranges are exactly what can cause troubles.

FYI there are first v6 dev builds available on NPM, and they might help with this too - I believe those dependency resolution troubles are partially caused by the peer dependency definitions we have, and those were removed in v6. 

https://github.com/mikro-orm/mikro-orm/blob/v6/docs/docs/upgrading-v5-to-v6.md

Would be interesting to verify if that behaves differently.

**Reply by zgid123** [reply]
cool. Lemme test this tonight

**Reply by zgid123** [reply]
I confirmed the new version (I think the version 5.6 can work too) can work well with NestJS. But I don't know why the new project can work, but my current side project cannot work. I am trying to reproduce for the test project

**Reply by zgid123** [reply]
two projects share the same config actually

**Reply by B4nan** [reply]
This has nothing to do with the ORM config, it is a dependency resolution issue - one of the packages is installed more than once, so if something, it can be caused by either the package manager you use or the way the ORM defines the (peer) dependencies - and that is why I suggested trying v6 dev builds, as those don't define the peer dependencies anymore.

**Reply by zgid123** [reply]
Yup. V6 works well when I created new project. I am trying to debug the reason. Thanks for your help.

This is off topic, I checked the current features, we still not support tree and table per type right? Are they good for this project and also is there any one work on this? I wanna research this library deeply and try to implement the `Tree entity` because one of my project need this technique. What do you think?

**Reply by B4nan** [reply]
I don't understand what you think is missing for "tree type". Entities can form a tree, you dont need anything special for that.

**Reply by zgid123** [reply]
I mean those techniques. If this is not good for this project, then I will make it as work around for my project

https://typeorm.io/tree-entities#closure-table

**Reply by B4nan** [reply]
You can do that here too, I am not planning to introduce any special support for that. It does not add much value, and it feels rather confusing to me - like why should you define your entities with some special decorators for such a basic use case? The only added value is in the addition find methods, and I would argue it can be rather confusing (e.g. from reading the code/docs I have no idea what queries this actually produces). You could define such methods in a custom repository too, they don't do anything special, just construct a query based on the closure table pattern.

If something, this can be born as a 3rd party project.

**Reply by zgid123** [reply]
I just tested again my test project. It seems whenever I added package `migrations`, it will causes the error

Here for the repo: https://github.com/zgid123/test-mikro-with-nestjs

**Reply by zgid123** [reply]
if I remove the package `migrations`, the injection will work

**Reply by zgid123** [reply]
package `seeder` is ok, package `cli` will throw error like package `migrations`

**Reply by zgid123** [reply]
I already tried `rm -rf node_modules` for root folder and workspace folder, even the `pnpm-lock.yaml` and reinstall all the packages

**Reply by B4nan** [reply]
I see, I will need to remove the peer dependencies there too I guess, they shouldn't be needed in v6, except the core package.

**Reply by zgid123** [reply]
Thank you so much. And thanks for creating this awesome package. I really like this package

**Reply by B4nan** [reply]
5.1.3 should help with this, I found a better way to register the EM flavor.

**Reply by zgid123** [reply]
cool. I will try it. Thanks for the update

**Comment by zgid123** [comment, marked as answer]
This problem is solved. I tested with `@mikro-orm` version `6.0.0-dev.15` and `@mikro-orm/nestjs` `5.1.3`

**Comment by zgid123** [comment]
sorry for bothering you again. I tried to debug but couldn't find out the issue or solution.

https://github.com/zgid123/test-mikro-with-nestjs/blob/main/workspaces/api-gateway/src/__tests__/specs/tests/tests.controller.spec.ts#L18

I created a package to provide a helper method to init all shared modules that some projects in monorepo will use it. But NestJS always raise 

```sh
    Nest can't resolve dependencies of the MikroOrmCoreModule (Symbol(mikro-orm-module-options), ?). Please make sure that the argument ModuleRef at index [1] is available in the MikroOrmCoreModule context.

    Potential solutions:
    - Is MikroOrmCoreModule a valid NestJS module?
    - If ModuleRef is a provider, is it part of the current MikroOrmCoreModule?
    - If ModuleRef is exported from a separate @Module, is that module imported within MikroOrmCoreModule?
      @Module({
        imports: [ /* the Module containing ModuleRef */ ]
      })
```

But if I use the `Test.createTestingModule` directly for the project, it does not raise error.

Old version didn't throw this error as I remember

**Reply by zgid123** [reply]
@B4nan sorry for bothering you again. I just tested for my side project. If I use latest version of this package, when I install specific 5.1.5 to both libs/testing and my api gateway, if the api gateway can run, the unit test will raise error above and vice versa.

If I install old package (5.1.2), I can make both works.

For version 5.1.5, I tried to pass the `MikroOrmModule` as reference to the helper `createModule`, but it didn't work.

I already tried to clean the node_modules and reinstall it

**Reply by Cryxto** [reply]
well i encounter this issue on version 6 also

**Reply by B4nan** [reply]
are you using `useFactory` option with injected dependencies? if yes, see https://github.com/mikro-orm/nestjs#driver-specific-imports

**Reply by Cryxto** [reply]
hello thx for reply and solution, i finally fix it by upgrade to 6.1 or above (i don't know why when install using pnpm it is default to 6 ) also i follow this discussion [https://github.com/mikro-orm/nestjs/discussions/184](https://github.com/mikro-orm/nestjs/discussions/184) which is like in the docs you mention . I have suggestion #cmiiw to not only contextName that should be omitted from useFactory return type but also the  driver option.
